consensus/src/transaction_shuffler/fairness/conflict_zone.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction_shuffler::fairness::{
+    conflict_key::{ConflictKeyId, ConflictKeyRegistry, MapByKeyId},
+    TxnIdx, NUM_CONFLICT_ZONES,
+};
+use std::collections::VecDeque;
+
+#[derive(Debug)]
+pub(crate) struct ConflictZone<'a> {
+    sliding_window_size: usize,
+    sliding_window: VecDeque<ConflictKeyId>,
+    counts_by_id: MapByKeyId<usize>,
+    key_registry: &'a ConflictKeyRegistry,
+}
+
+impl<'a> ConflictZone<'a> {
+    pub fn build_zones(
+        key_registries: &'a [ConflictKeyRegistry; NUM_CONFLICT_ZONES],
+        window_sizes: [usize; NUM_CONFLICT_ZONES],
+    ) -> [Self; NUM_CONFLICT_ZONES] {
+        itertools::zip_eq(key_registries.iter(), window_sizes)
+            .map(|(registry, window_size)| Self::new(registry, window_size))
+            .collect::<Vec<_>>()
+            .try_into()
+            .expect("Impossible to fail.")
+    }
+
+    fn new(key_registry: &'a ConflictKeyRegistry, sliding_window_size: usize) -> Self {
+        Self {
+            sliding_window_size,
+            sliding_window: VecDeque::with_capacity(sliding_window_size + 1),
+            counts_by_id: key_registry.new_map_by_id(),
+            key_registry,
+        }
+    }
+
+    pub fn is_conflict(&self, txn_idx: TxnIdx) -> bool {
+        let key_id = self.key_registry.key_id_for_txn(txn_idx);
+        if self.key_registry.is_conflict_exempt(key_id) {
+            false
+        } else {
+            *self.counts_by_id.get(key_id) > 0
+        }
+    }
+
+    // Append a new transaction to the sliding window and
+    // return the key_id that's no longer in conflict as a result if there is one.
+    pub fn add(&mut self, txn_idx: TxnIdx) -> Option<ConflictKeyId> {
+        let key_id = self.key_registry.key_id_for_txn(txn_idx);
+
+        *self.counts_by_id.get_mut(key_id) += 1;
+        self.sliding_window.push_back(key_id);
+        if self.sliding_window.len() > self.sliding_window_size {
+            let removed_key_id = self.sliding_window.pop_front().unwrap();
+            let count = self.counts_by_id.get_mut(removed_key_id);
+            *count -= 1;
+            if *count == 0 && !self.key_registry.is_conflict_exempt(removed_key_id) {
```

Comment:
May be I am missing something but why do we need to check if the key is not conflict exempt? Irrespective of it, if the count is 0, we can return this as this has no conflict in the window?

Reply:
we return what's UN-conflicted, for which the conflict zones and pending zones will be checked in a loop -- we will save by not returning those exempted.